mcs/class/System/Mono.AppleTls/AppleTlsContext.cs

```diff
@@ -0,0 +1,931 @@
+			int i = 0;
+
+			int n = 0;
+			if (certificates != null)
+			{
+				using (var enumerator = certificates.GetEnumerator())
+    					while (enumerator.MoveNext())
+        					n++;
+			}
+
+			var ptrs = new IntPtr [n + 1];
+			ptrs [0] = identity.Handle;
+			foreach (var certificate in certificates)
+				ptrs [++i] = certificate.Handle;
+			return CFArray.CreateArray (ptrs);
+		}
+
+		public void SetCertificate (SecIdentity identify, IEnumerable<SecCertificate> certificates)
+		{
+			using (var array = Bundle (identify, certificates)) {
+				var result = SSLSetCertificate (Handle, array.Handle);
+				CheckStatusAndThrow (result);
+			}
+		}
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern unsafe static /* OSStatus */ SslStatus SSLGetClientCertificateState (/* SSLContextRef */ IntPtr context, out SslClientCertificateState clientState);
+
+		public SslClientCertificateState ClientCertificateState {
+			get {
+				SslClientCertificateState value;
+				var result = SSLGetClientCertificateState (Handle, out value);
+				CheckStatusAndThrow (result);
+				return value;
+			}
+		}
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern unsafe static /* OSStatus */ SslStatus SSLCopyPeerTrust (/* SSLContextRef */ IntPtr context, /* SecTrustRef */ out IntPtr trust);
+
+		public SecTrust GetPeerTrust (bool requireTrust)
+		{
+			IntPtr value;
+			var result = SSLCopyPeerTrust (Handle, out value);
+			if (requireTrust) {
+				CheckStatusAndThrow (result);
+				if (value == IntPtr.Zero)
+					throw new TlsException (AlertDescription.CertificateUnknown);
+			}
+			return (value == IntPtr.Zero) ? null : new SecTrust (value);
+		}
+
+		#endregion
+
+		#region IO Functions
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern static /* SSLContextRef */ IntPtr SSLCreateContext (/* CFAllocatorRef */ IntPtr alloc, SslProtocolSide protocolSide, SslConnectionType connectionType);
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern static /* OSStatus */ SslStatus SSLSetConnection (/* SSLContextRef */ IntPtr context, /* SSLConnectionRef */ IntPtr connection);
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern static /* OSStatus */ SslStatus SSLSetIOFuncs (/* SSLContextRef */ IntPtr context, /* SSLReadFunc */ SslReadFunc readFunc, /* SSLWriteFunc */ SslWriteFunc writeFunc);
+
+		[MonoPInvokeCallback (typeof (SslReadFunc))]
+		static SslStatus NativeReadCallback (IntPtr ptr, IntPtr data, ref /*HACK*/ IntPtr dataLength)
+		{
+			var handle = GCHandle.FromIntPtr (ptr);
+			if (!handle.IsAllocated)
+				return SslStatus.Internal;
+
+			var context = (AppleTlsContext) handle.Target;
+			if (context.disposed)
+				return SslStatus.ClosedAbort;
+
+			try {
+				return context.NativeReadCallback (data, ref dataLength);
+			} catch (Exception ex) {
+				if (context.lastException == null)
```

Comment:
Be very careful with modifying the exception handling code, especially when it comes to making large structural changes as you can quickly get yourself into very big trouble.  Changes like those should not be done when simply moving code from one repository to another, but on their own - and only after careful evaluation and testing.

I'm not very familiar with what `ExceptionDispatchInfo.Capture` does - but the words "capture" and "rethrow" raise big red flags in a managed -> native -> managed scenario.